node/silkworm/downloader/internals/priority_queue.hpp

```diff
@@ -26,7 +26,8 @@
  */
 template <typename T, typename CMP>
 class set_based_priority_queue {
-    std::multiset<T, CMP> elements_;
+    using impl_t = std::multiset<T, CMP>;
```

Comment:
if performance is a concern, `absl::btree_multiset`  would probably be significantly faster.

Reply:
Thanks! Actually I need a queue of Links that:
a) it is ordered by block_number => for headers purge task
b) support search by (block_number, hash) pair => for headers verification task

So I am thinking to use one of the following:

1) a multimap<BlockNum,Link> (ordered by BlockNum)
where 
- a) is naturally implemented
- b) is implemented making an equal_range(BlockNum) operations and than searching for the specific hash

or 

2) a boost multi-index container with 2 indexes, one for BlockNum ordering and one for (block_number, hash) searching

What do you think about?

Reply:
I think the easiest would be to use a `btree_set<std::pair<BlockNum, Hash>>`. This will be sorted first by BlockNum, then by the Hash (default lexicographical comparison by `std::pair`. Makes it trivial to do both tasks you mention. There is some duplication of BlockNum, but it is pretty small.

to purge all elements until block #x you would do something like:

```
using HSet = btree_set<std::pair<BlockNum, Hash>>;
HSet set;

set.erase(set.begin(), set.lower_bound(HSet::value_type(x,0)));
```

Reply:
Very interesting solution. I will record it for the time in which I will measure performances & optimize them.
Thanks!